src/approximate-qvm.lisp

```diff
@@ -0,0 +1,173 @@
+  "Given a value for t_phi (dephasing time) and a gate time, calculates the kraus
+operators corresponding to the dephasing noise."
+  (let*  ((prob (/ gate-time tphi))
+          (prob-k1 (/ prob 2))
+          (prob-k0 (- 1 prob-k1 ))
+          (kraus-ops (loop :for mat :in '("I" "Z")
+                           :for pj :in (list prob-k0 prob-k1)
+                           :collect (magicl:scale (sqrt pj)
+                                                  (quil:gate-matrix
+                                                   (quil:gate-definition-to-gate
+                                                    (quil:lookup-standard-gate mat)))))))
+    kraus-ops
+    )) 
+
+(defun generate-damping-dephasing-kraus-ops (t2 gate-time)  
+  "Calculates the kraus operators for dephasing and damping noise from t2, which
+is decoherence time. "
+  (generate-dephasing-kraus-ops t2 gate-time)) 
+
+
+(defun kraus-kron (k1s k2s)
+  "Calculate the tensor product of two kraus maps by tensoring their elems."
+  (loop :for x in k1s
+        :append (loop :for y in k2s
+                      :collect (magicl:kron x y))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defun simple-test-approx ()
+  (let* ((ones 0)
+         (qvm (make-instance 'approx-qvm :number-of-qubits 2 :avg-gate-time 1))
+         (numshots 100)
+         (program "DECLARE R0 BIT; X 0; CNOT 0 1;  MEASURE 0 R0")
+         (parsed-program  (quil:parse-quil program)) )
+    (load-program qvm parsed-program :supersede-memory-subsystem t)
+    (set-qubit-t1 qvm 0 5)
+    (set-qubit-t1 qvm 1 4)
+    (set-qubit-t2 qvm 0 3)
+    (set-qubit-t2 qvm 1 5)
+    (run qvm))
+  )
+
+(defun test-approx ()
+  "Given a value for t1 and a gate time, generates the kraus operators corresponding
+to the t1 noise. "
+  (let ((ones 0)
+        (qvm (make-instance 'approx-qvm :number-of-qubits 1 :t1 5 :avg-gate-time 1))
+        (numshots 100)
+        (program "DECLARE R0 BIT; X 0; MEASURE 0 R0"))
+    (loop :repeat numshots
+          :do (bring-to-zero-state (amplitudes qvm))
+          :do (incf ones (do-noisy-measurement qvm 0 program)))
+    (format t "result: ~a" (float (/ ones numshots)))))
+
+
+(defun test-approx-fro ()
+  "Given a value for t1 and a gate time, generates the kraus operators corresponding
+to the t1 noise. "
+  (let* ((ones 0)
+         (qubit 0)
+         (qvm (make-instance 'approx-qvm :number-of-qubits 1 :avg-gate-time 1))
+         (numshots 100)
+         (program "DECLARE R0 BIT; X 0;  MEASURE 0 R0"))
+    
+    (set-qubit-fro qvm qubit .9d0)
+    (loop :repeat numshots
+          :do (bring-to-zero-state (amplitudes qvm))
+          :do (incf ones (do-noisy-measurement qvm 0 program)))
+    (format t "result: ~a" (float (/ ones numshots)))))
+
+
+(defun fro-map (qubit fro)
+  (setf m (make-hash-table))
```

Comment:
Just a warning, Common Lisp has several notions of object equality, and hash tables allow for this choice with the `:test` keyword argument. The default is `'eql`, which is fine for integers. If you are keying `quil:qubit` objects this won't work FYI

Reply:
For now the qubit here is an int, but i'll keep this in mind if that changes